tasks/codex/edm.py

```diff
@@ -0,0 +1,528 @@
+
+        loc = edm_data.get('locationId')
+        if loc:
+            found = reg.transfer_find(loc)
+            if found and found.get('id'):
+                sighting.location_guid = found['id']
+            else:  # TODO handle better?
+                raise ValueError(
+                    f'unknown locationId "{loc}" on sighting {sighting.guid}'
+                )
+
+        txs = edm_data.get('taxonomies', [])
+        taxonomies = []
+        for tx_id in txs:
+            # will raise ValueError if bad id
+            taxonomies.append(Taxonomy(tx_id))
+        if taxonomies:
+            sighting.set_taxonomies(taxonomies)
+
+        edm_custom_fields = edm_data.get('customFields', {})
+        sighting.set_custom_field_values(edm_custom_fields)
+    db.session.flush()
+    print('sighting edm data transfer complete')
+
+
+def transfer_data_individual():
+    import datetime
+
+    import app.modules.utils as util
+    from app.modules.individuals.models import Individual
+    from app.modules.site_settings.models import Taxonomy
+
+    indivs = Individual.query.all()
+    print('individual edm data transfer started')
+    for indiv in tqdm.tqdm(indivs):
+        response = app.edm.get_dict('individual.data_complete', indiv.guid)
+        assert isinstance(response, dict)
+        assert response.get('success', False)
+        edm_data = response['result']
+        sex = edm_data.get('sex')
+        if sex and not util.is_valid_sex(sex):
+            raise ValueError(f'invalid sex "{sex}" on individual {indiv.guid}')
+        indiv.sex = sex
+        indiv.comments = edm_data.get(
+            'comments'
+        )  # TODO are comments staying here or customField via migration? FIXME
+
+        tx_id = edm_data.get('taxonomy')
+        if tx_id:
+            Taxonomy(tx_id)  # will raise ValueError if bad id
+            indiv.taxonomy_guid = tx_id
+
+        tob = edm_data.get('timeOfBirth')
+        if tob and tob != '0':
+            try:
+                # i am not sure if the value is in ms or sec?  FIXME
+                indiv.time_of_birth = datetime.datetime.fromtimestamp(float(tob) / 1000.0)
+            except Exception:
+                log.warning(f'could not get datetime from "{tob}"')
+        tod = edm_data.get('timeOfDeath')
+        if tod and tod != '0':
+            try:
+                indiv.time_of_death = datetime.datetime.fromtimestamp(float(tod) / 1000.0)
+            except Exception:
+                log.warning(f'could not get datetime from "{tod}"')
+
+        edm_custom_fields = edm_data.get('customFields', {})
+        indiv.set_custom_field_values(edm_custom_fields)
+    db.session.flush()
+    print('individual edm data transfer complete')
+
+
+@app_context_task(
+    help={
+        'section': 'setting, user, individual, encounter, sighting (note: setting MUST be run before any others will work)',
+    }
+)
+def transfer_data(context, section=None, refresh=True):
```

Comment:
Just wondering if this needs some "Are you really sure that you want to be doing this, this could really scribble all over your existing data and should not be carried out other than as part of transitioning data from edm as a one time operation"

Reply:
hmmm yeah, its a reasonable question.   i am trying to think of a clever way to detect "has this ever happened before?" to maybe make it not so harsh _the first time_.  but even then, i think thats dangerous.

there _is_ a kind of built in safety mechanism worth noting, however:
```
    if not is_extension_enabled('edm'):
        raise RuntimeError('EDM must be enabled')
```
basically a user _cannot_ run this task on a "normal" running codex -- as it wont have edm. in fact, i would argue its pretty painful to get edm going, so would likely not be accidental.